File: Septimo_Cuatrimestre/SO/Prácticas/P3/practica3/03-cpu-bound/klt.c
#define _POSIX_C_SOURCE 200809L
#include <pthread.h>
#include <stdio.h>
#include <unistd.h>
#include "common.h" // clockwatch_t, difftime_ns, nth_prime_number

#define SLEEP_TIME 10
#define NUM_THREADS 5

// Count the number of primes up to this limit
#define LIMIT 2500000


void* do_work(void* arg) {
    pthread_t thread_id = pthread_self();
    clockwatch_t clock;
    double elapsed;

    
    printf("[Thread %lu] Doing some work...\n", thread_id);
    clockwatch_start(&clock);

    // Calculate the nth prime number
    nth_prime_number(LIMIT); // CPU bound work
    
    elapsed = clockwatch_stop(&clock);
    printf("[Thread %lu] Done with work in %f seconds.\n", thread_id, elapsed);
    return NULL;
}

int main() {
    pthread_t threads[NUM_THREADS];
    clockwatch_t clock;
    double elapsed;

    clockwatch_start(&clock);
    printf("Starting the program.\n");

    for (int i = 0; i < NUM_THREADS; i++) {
        pthread_create(&threads[i], NULL, do_work, NULL);
    }

    for (int i = 0; i < NUM_THREADS; i++) {
        pthread_join(threads[i], NULL);
    }

    elapsed = clockwatch_stop(&clock);
    printf("All threads are done in %f seconds\n", elapsed);

    return 0;
}

File: Cuarto/SO/Prácticas/P3/practica3/03-cpu-bound/klt.py
import threading
import time
from common import nth_prime_number

# Number of threads to spawn
NUM_THREADS = 7

# Delay in seconds to simulate IO-bound work
# adjust for faster results or to have extra time to monitor resources
SLEEP_TIME = 10

# Count the number of primes up to this limit
LIMIT = 500_000


def do_work():
    thread_id = threading.get_ident()
    start = time.time()
    print(f"[{thread_id=}] Doing some work...")
    nth_prime_number(LIMIT)  # This is a CPU-bound task
    elapsed = time.time() - start
    print(f"[{thread_id=}] Done with work in {elapsed} seconds.")


if __name__ == "__main__":
    start = time.time()
    print("Starting the program.")
    all_threads = []
    for i in range(NUM_THREADS):
        # Start NUM_THREADS threads
        thread = threading.Thread(target=do_work)
        all_threads.append(thread)
        thread.start()

    for thread in all_threads:
        # Wait for all threads to finish
        thread.join()

    elapsed = time.time() - start
    print(f"All threads are done in {elapsed} seconds")
